src/choose-location/choose-location.component.tsx

```diff
@@ -0,0 +1,118 @@
+import React from "react";
+import { always } from "kremling";
+import { useConfig } from "@openmrs/esm-module-config";
+import { Trans } from "react-i18next";
+import {
+  getLoginLocations,
+  setSessionLocation
+} from "./choose-location.resource";
+import styles from "../styles.css";
+
+export default function ChooseLocation(props: ChooseLocationProps) {
+  const config = useConfig();
+  const [loginLocations, setLoginLocations]: [
+    Array<Object>,
+    (locations: Array<Object>) => void
+  ] = React.useState([]);
+  const [location, setLocation] = React.useState("");
+  const [isSubmitting, setIsSubmitting] = React.useState(false);
+  const [errorMessage, setErrorMessage] = React.useState("");
+  const locationInputRef = React.useRef<HTMLInputElement>(null);
+  const formRef = React.useRef<HTMLFormElement>(null);
+
+  React.useEffect(() => {
+    const sub = getLoginLocations().subscribe(locations => {
+      setLoginLocations(locations);
+    });
+
+    return () => sub.unsubscribe();
+  }, []);
+
+  React.useEffect(() => {
+    if (isSubmitting) {
+      setSessionLocation(location).then(data => {
+        navigate(props, config.links.loginSuccess);
+      });
+    }
+  }, [isSubmitting]);
+
+  const CardHeader = props => (
```

Comment:
Inlining components inside of another component is bad for performance and can result in nasty bugs such as unexpected loss of focus, multiple network requests when only one was intended, etc. The reason is that the React reconciler checks the [react element's type](https://reactjs.org/docs/jsx-in-depth.html#specifying-the-react-element-type) every render and does an `oldType === newType` check. If that check returns false, it **unmounts the old component and mounts the new component**. You can read more about this [in these React docs](https://reactjs.org/docs/reconciliation.html#elements-of-different-types)

In your case here, you are (unintentionally) recreating the CardHeader and RadioInput components every time that the ChooseLocation is re-rendered. The reason is that the CardHeader function is inside of the ChooseLocation function and so it has to be recreated every render so that it has access to the new props ("javascript closure"). This means that the components (and I think the dom elements) will be torn down and recreated every render.

To avoid this, you have two options:

Option 1 (my favorite way) - use inner functions that aren't react components
```js
function Foo(props) {
  return (
    <div>{renderThing()}</div>
  )

  function renderThing() {
    return props.thing
  }
}
```

Option 2: move Card Header and Radio Input outside of the ChooseLocation function.

Reply:
A component that uses `props.children` would seem to call for Option 2?

Reply:
No, it doesn't. A function that is not a component has access to the exact same props and state. It also can be passed JSX as an argument

Reply:
Is CardHeader already a non-component function as written? Just a function that takes (props) as argument and returns some JSX?

Reply:
CardHeader, as written right now, is a component. See line 67:

https://github.com/openmrs/openmrs-esm-login/blob/1b49aa156e95f44068fc796b11d29ec836e97230/src/choose-location/choose-location.component.tsx#L67

See [this babel repl](https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=DwYQhgTgJgEgpmKcIAID0A-IA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=es2015%2Creact%2Cstage-1%2Cstage-2&prettier=false&targets=&version=7.7.4&externalPlugins=) that shows that this gets compiled to the following:

```jsx
React.createElement(CardHeader)
```

Now see [the react docs for createElement](https://reactjs.org/docs/react-api.html#createelement) which show that the first argument to createElement is the `type`. Passing a function in as the `type` means that that function is a React component. Meanwhile, compare to the following example, where the function is **invoked before even returned to React as a react element**:

```jsx
function Foo() {
  return <div>{thing()}</div>

  function thing() {
    return <button>thing</button>
  }
}
```

Note how the `thing` function is not given a props object since it's not a component. Also note that it is allowed to be named with a lowercase letter at the first letter, [even though JSX  requires that components are written with a capital letter](https://reactjs.org/docs/glossary.html#components). The reason why is because we're not returning `thing` as a react element `type`. From react's perspective, our Foo function just returned a button inside of a div.